src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentremoval/DocumentRemovalService.java

```diff
@@ -49,4 +162,89 @@ public void retrieveDocumentNodes(JsonNode root, List<JsonNode> documentNodes) {
+
+            while (fieldNames.hasNext()) {
+                String fieldName = fieldNames.next();
+                JsonNode fieldValue = root.get(fieldName);
+
+                if (shouldRemoveDocument(fieldValue,
+                    documentToDelete.getCaseDocument().getDocumentUrl())) {
+                    log.info(String.format("Deleting doc with url %s", documentToDelete.getCaseDocument().getDocumentUrl()));
+                    fieldsToRemove.add(fieldName);
+                } else {
+                    removeDocumentFromJson(fieldValue, documentToDelete);
+                }
+            }
+        } else if (root.isArray()) {
+            processArrayNode(root, documentToDelete);
+        }
+
+        for (String fieldName : fieldsToRemove) {
+            ((ObjectNode) root).remove(fieldName);
+        }
+    }
+
+    private void processArrayNode(JsonNode root, DocumentToKeep documentToDelete) {
+        ArrayNode arrayNode = (ArrayNode) root;
+        for (int i = 0; i < arrayNode.size(); i++) {
+            JsonNode arrayElement = arrayNode.get(i);
+            if (arrayElement.has(VALUE_KEY)) {
+                JsonNode valueObject = arrayElement.get(VALUE_KEY);
+                Iterator<String> fieldNames = valueObject.fieldNames();
+
+                while (fieldNames.hasNext()) {
+                    String fieldName = fieldNames.next();
+                    JsonNode fieldValue = valueObject.get(fieldName);
+
+                    if (shouldRemoveDocument(fieldValue,
+                        documentToDelete.getCaseDocument().getDocumentUrl())) {
+                        log.info(String.format("Deleting doc with url %s", documentToDelete.getCaseDocument().getDocumentUrl()));
+                        ((ArrayNode) root).remove(i);
+                    }
+                }
+            }
+            removeDocumentFromJson(arrayElement, documentToDelete);
+        }
+    }
+
+    private boolean shouldRemoveDocument(JsonNode fieldValue, String documentToKeepUrl) {
+        return fieldValue.has(DOCUMENT_URL)
+                && fieldValue.get(DOCUMENT_URL).asText().equals(documentToKeepUrl);
+    }
+
+    // Consider making async again.  See deleteOldMiniFormA
+    private void deleteDocument(DocumentToKeep documentToRemove, String authorisationToken) {
+        try {
+            log.info(String.format("Deleting doc from DocStore with url %s",
+                documentToRemove.getCaseDocument().getDocumentUrl()));
+
+            if (featureToggleService.isSecureDocEnabled()) {
+                genericDocumentService.deleteDocument(documentToRemove.getCaseDocument().getDocumentUrl(), authorisationToken);
+            }
+        } catch (Exception e) {
+            log.error(format(
+                    "Failed to delete document url %s",
+                    documentToRemove.getCaseDocument().getDocumentUrl()), e);
+
+            throw new DocumentDeleteException(e.getMessage(), e);
+        }
+    }
```

Comment:
Ideally leave this comment.  The async approach hasn't been tested in demo, but the the sync approach has.
When isSecureDocEnabled is toggled on, files will begin deleting upon file removal.  There would be an option to make file deletion async then for performance.